addons/product/models/product_template.py

```diff
@@ -938,6 +939,61 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, attribute_values):
+        """
+        Generate all possible combination for attributes values.
+        It is equivalent to itertools.product except it skips invalid combinations.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+        """
+        if not attribute_values:
+            raise StopIteration
+
+        pools = attribute_values
+
+        all_exclusions = self._get_own_attribute_exclusions()
```

Comment:
This is actually a good suggestion, indeed it seems like maybe the true end goal here is to be able to not call `_is_combination_possible` at all after the process is done?

Reply:
The only problem is that duplicate code is not good for maintenance, and if it's done incorrectly here, all the test that cover the other method will not catch it.

Reply:
## Tests
The method `get_possible_combinations` is thoroughly tested, and its scope hasn't changed at all.
So if the tests were good before my code, they are still good now.
The only change is performance wise and that's why I added a performance test.

## DRY
Trying to factorize at all cost is also bad.

Now, the new code only "duplicates" the following part of the `_is_combination_possible` :
```python
        exclusions = self._get_own_attribute_exclusions()
        if exclusions:
            # exclude if the current value is in an exclusion,
            # and the value excluding it is also in the combination
            for ptav in combination:
                for exclusion in exclusions.get(ptav.id):
                    if exclusion in combination.ids:
                        return False
```
It does so because it needs to test a **partial** combination rather than a full combination. So it doesn't do exactly the same thing to begin with.  Even less now that I've further modified the code (see code, I've missed the case where a value is incompatible with a previous value already included in the partial combination).
I could indeed include parent exclusions (done), it's a good idea as it could prevent performance issues in some cases, I think.
But what about the parts before ?
The method `_is_combination_possible` is called at 4 places, 2 of which outside the method I'm modifying. So at the very least, the method `_is_combination_possible` cannot cease to exist.

So, I could try to factorize this code, so it's not "duplicated" but I think that'll make the code more obscure and not clearer.

Reply:
I don't think we need to factorize the exclusion part, especially the parent exclusion was trivial to handle in your method, which was good to see.